gio-subclass/src/action.rs

```diff
@@ -0,0 +1,202 @@
+
+    fn get_state_type(&self, action: &Action) -> Option<glib::VariantType>;
+
+}
+
+any_impl!(ActionImpl);
+
+pub trait ActionImplStatic<T: ObjectType>: 'static {
+    fn get_impl<'a>(&self, imp: &'a T::ImplType) -> &'a ActionImpl;
+}
+
+struct ActionStatic<T: ObjectType>{
+    imp_static: *const ActionImplStatic<T>
+}
+
+unsafe extern "C" fn action_activate<T: ObjectType>
+(ptr: *mut GAction, parameter: *mut glib::GVariant)
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let klass = &**(ptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(ffi::action_get_type())
+                                     as *const ActionStatic<T>;
+    let instance = &*(ptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap: T = from_glib_borrow(instance);
+    imp.activate(&wrap, &from_glib_none(action), /*Unknown conversion*/parameter)
+}
+
+unsafe extern "C" fn action_change_state<T: ObjectType>
+(ptr: *mut GAction, value: *mut glib::GVariant)
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let klass = &**(ptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(ffi::action_get_type())
+                                     as *const ActionStatic<T>;
+    let instance = &*(ptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap: T = from_glib_borrow(instance);
+    imp.change_state(&wrap, &from_glib_none(action), &from_glib_none(value))
+}
+
+unsafe extern "C" fn action_get_enabled<T: ObjectType>
+(ptr: *mut GAction) -> gboolean
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let klass = &**(ptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(ffi::action_get_type())
+                                     as *const ActionStatic<T>;
+    let instance = &*(ptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap: T = from_glib_borrow(instance);
+    imp.get_enabled(&wrap, &from_glib_none(action)).to_glib()
+}
+
+unsafe extern "C" fn action_get_name<T: ObjectType>
+(ptr: *mut GAction) -> *const c_char
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let klass = &**(ptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(ffi::action_get_type())
+                                     as *const ActionStatic<T>;
+    let instance = &*(ptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap: T = from_glib_borrow(instance);
+    imp.get_name(&wrap, &from_glib_none(action))/*Not checked*/.to_glib_none().0
```

Comment:
The return value is unsafe but that's because the C API is broken, this needs a manual implementation.

You need to return a string here that is valid for the whole lifetime of the underlying action object

Reply:
@federicomenaquintero things like this are btw examples why I think we will need an override system somewhere between gnome-class and the .gir files, and why I believe the trampolines and some glue infrastructure should live in the gtk/etc bindings for gnome-class to use.

I have some more examples in GStreamer :) We should talk about this at the next hackfest in more detail